Restoring Append-only security requirements (#212)
ietf-wg-scitt/draft-ietf-scitt-architecture: draft-ietf-scitt-architecture.md

  I-D.draft-ietf-cose-merkle-tree-proofs: COMETRE

  RFC2397: DataURLs

>   RFC6024:

Reviewer: This informative RFC is about trust anchor management and that is material to the changes being mad, did you or someone reviewing intend to remove it?

Sorry suggestions does not support adding back deleted lines anymore or I would just recommend it.

Reviewer 2: Trust Anchor Management is a vital part of SCITT story. I suggest we should retain it, perhaps needs to be referred from the section that talks about setting the Transparency Service with necessary Trust Anchors!

Reviewer 3: Good catch. I removed the reference as it was no longer in the doc, and called out as a make error:

The reference to RFC6024 was removed in [PR 201](https://github.com/ietf-wg-scitt/draft-ietf-scitt-architecture/pull/201/files#diff-585fb424519b79cf000445f3425fa56d328cbaca5d2a622740ddc4e5ca91dbe1L394)

Non-equivocation:

: a state where it is impossible for a Transparency Service to provide different views of its Append-only Log to Verifiers about the same Artifact.
Over time, an Issuer may register new Signed Statements about an Artifact in a Transparency Service with new information. However, the consistency of a collection of Signed Statements about the Artifact can be checked by all Verifiers.

Receipt:

: a cryptographic proof that a Signed Statement is included in the Append-only Log.

> Receipts are based on Signed Inclusion Proofs as described in COSE Signed Merkle Tree Proofs {{-COMETRE}};

Reviewer 4: ```suggestion
Receipts are based on Signed Inclusion Proofs, such as those as described in COSE Signed Merkle Tree Proofs {{-COMETRE}};
```

> they can be built on different verifiable data structures, not just binary merkle trees.

Reviewer: Isn't this the only one we have defined? Why not put them in a table and make it small changes to add if there is presumably more than one? It seems we have COMETRE and the CCF profile for COMETER, so I am not sure why we would want to emphasize this when no public draft or IETF spec drafts an alternative one.

Am I missing something?

Reviewer 3: While COMETRE is currently the only one defined, the spec allows for more. Not sure the spec needs to list all the current and potential future implementations. 
To get this merged, I'm deferring this to a potential new PR, if felt needed.

Receipt:

: a cryptographic proof that a Signed Statement is included in the Append-only Log.
Receipts are based on Signed Inclusion Proofs as described in COSE Signed Merkle Tree Proofs {{-COMETRE}};
they can be built on different verifiable data structures, not just binary merkle trees.

> A receipts consists of an Transparency Service-specific inclusion proof for the Signed Statement, a signature by the Transparency Service of the state of the Append-only Log after the inclusion, and additional metadata (contained in the signature's protected headers) to assist in auditing.

Reviewer 2: ```suggestion
A receipt consists of an Transparency Service-specific inclusion proof for the Signed Statement, a signature by the Transparency Service of the state of the Append-only Log after the inclusion, and additional metadata (contained in the signature's protected headers) to assist in auditing.
```

> A Receipt consists of an Transparency Service-specific inclusion proof for the Signed Statement, a signature by the Transparency Service of the state of the Append-only Log after the inclusion, and additional metadata (contained in the signature's protected headers) to assist in auditing.

Reviewer 5: ```suggestion
A Receipt consists of a Transparency Service-specific inclusion proof for the Signed Statement, a signature by the Transparency Service of the state of the Append-only Log after the inclusion, and additional metadata (contained in the signature's protected headers) to assist in auditing.
```

## Transparency Service

Transparency Services MUST feature an Append-only Log.

> The Append-only Log is a verifiable data structure that accumulates all Signed Statements 

Reviewer 6: I wonder if the Log is allowed to accumulate anything OTHER than "Signed Statements"... for example certificates, or JWTs.

Reviewer 3: I preferred the previous version as it incorporates generating receipts. This also changes registered to accumulates, which only adds more concepts.

Reviewer 7: Allowing the TS to use the log for audit information, in particular intermediate certificates or JWTs, is a pretty legitimate use case, so I would say yes

Reviewer: So we are saying, @ad-l we MUST add alternatives data to Signed Statements? SHOULD?

This has been discussed a few times in calls and on the mailing list, why MUST or SHOULD it be in the log and not something computed and returned by the API? I do not oppose that change, but why open the door?

And you want the certs and JWTs themselves in the log, @ad-l? I can understand signed or transparent statements if properly sealed, but this reads as using the log for generic storage. Beyond my personal opinion, I thought many in the group oppose this notion generally.

Reviewer 4: We need to distinguish between the log and the verifiable data structure if we are doing this. A merkle based append only log is just a bunch of hashes. The signed statements are committed to it, not "in it". I strongly prefer the original wording.

Reviewer 3: Marking as resolved: Please see the files view for the most recent changes.

registered by the Transparency Service and that supports the production of Receipts.

> 

Reviewer 7: ```suggestion
Optionally, the Ledger may also record other collateral information for Auditors, such as governance information, records of management operations or security incidents, intermediate certificates used for Issuer authentication, etc.

```

Reviewer: If it is just a different kind of payload, why must we be explicit with this?

Reviewer 3: Various implementations will support many different scenarios. I'd suggest change to 375/376 should be reverted, and remove the additional 378

Reviewer 4: The original wording here is much more preferable. The verifiabiliy depends on the ability to produce proofs (receipts) not where the signed statements are stored.

Reviewer 4: This gives me something extra to think about, sounds like a generaly useful thing, but doesn't seem necessary architecture ?

Reviewer 3: Resolving as not-needed, and implementation specific details.
@ad-l, please re-open with comments and changes if you feel we need to keep it

## Transparency Service

Transparency Services MUST feature an Append-only Log.

> The Append-only Log is a verifiable data structure that accumulates all Signed Statements registered by the Transparency Service and that supports the production of Receipts.

Reviewer 3: Proposal to revert this change, as "accumulates", as we've moved away from the Transparency Service recording Signed Statement contgents, as that may introduce PII or other redaction scenarios that would be impossible. Rather the Transparency Service registers signed statements, producing receipts.
```suggestion
The Append-only Log is the verifiable data structure that records Signed Statements and supports the production of Receipts.
```

Reviewer 2: I agree with @SteveLasker assessment and I would prefer the earlier `old` version of the text.

In fact I would re-phrase it to:
`An Append-only Log is the verifiable data structure that registers Signed Statements and produces Receipts.`

### Append-only Log

The security properties of the Append-only Log are determined by the choice of the verifiable data structure used by the Transparency Service to produce Receipts.
This verifiable data structure MUST support the following security requirements:

Append-Only:

: once included in the verifiable data structure, a Signed Statement cannot be modified, deleted, or reordered; hence its Receipt provides a universally-verifiable proof of registration.

Non-equivocation:

> : there is no fork in the Append-only Log. Everyone with access to its content sees the same collection of Signed Statements, and can check that it is consistent with any Receipts they have verified.

Reviewer 3: ```suggestion
: there is no fork in the Append-only Log.
Everyone with access to its content sees the same collection of Signed Statements and can check that it is consistent with any Receipts they have verified.
```

### Append-only Log

> The security properties of the Append-only Log are determined by the choice of the verifiable data structure used by the Transparency Service to produce Receipts.

Reviewer 2: ```suggestion
The security properties of the Append-only Log are determined by the choice of the verifiable data structure used by the Transparency Service to implement the Log.
```

This verifiable data structure MUST support the following security requirements:

Append-Only:

> : once included in the verifiable data structure, a Signed Statement cannot be modified, deleted, or reordered; hence its Receipt provides a universally-verifiable proof of registration.

Reviewer 3: ```suggestion
: once included in the verifiable data structure, a Signed Statement cannot be modified, deleted, or reordered; hence its Receipt provides an offline verifiable proof of registration.
```

Non-equivocation:

: there is no fork in the Append-only Log.
Everyone with access to its content sees the same collection of Signed Statements and can check that it is consistent with any Receipts they have verified.

Replayability:

> : the Append-only Log includes sufficient information to enable everyone with access to its content to check that each included Signed Statement has been correctly registered.

Reviewer 2: ```suggestion
: the Append-only Log includes sufficient information to enable authorised actors with access to its content to check that each included Signed Statement has been correctly registered.
```

### Append-only Log

The security properties of the Append-only Log are determined by the choice of the verifiable data structure used by the Transparency Service to produce Receipts. 
This verifiable data structure MUST support the following security requirements:

Append-Only: 

> : once included in the verifiable data structure, a Signed Statement cannot be modified, deleted, or reordered; hence its Receipt provides a universally-verifiable proof of registration. 

Reviewer 4: ```suggestion
once included in the verifiable data structure, a Signed Statement cannot be modified, deleted, or reordered without invalidating the cryptographic proof backing its Receipt; hence its Receipt provides an offline verifiable proof of tamper evident registration.
```

### Append-only Log

The security properties of the Append-only Log are determined by the choice of the verifiable data structure used by the Transparency Service to produce Receipts. 
This verifiable data structure MUST support the following security requirements:

Append-Only: 

> : once included in the verifiable data structure, a Signed Statement cannot be modified, deleted, or reordered; hence its Receipt provides a universally-verifiable proof of registration. 

Reviewer 6: I can imagine readers might get hung up on "universally-verifiable proof of registration".

I wonder if we should define SCITT receipts as "proof of registration" that build on COSE receipts, which are signed inclusion proofs.

Reviewer 5: I like 'offline verifiable' a lot more than 'universally'

Reviewer 3: > I like 'offline verifiable' a lot more than 'universally'

Suggested change added

Reviewer 4: This just isn't true tho is it ? If a service has a take down request or whatever, it is legaly obliged to destroy the statement. This is usually accomplished by deleting the pre-image or "forgeting" some encrypting key

Reviewer 4: I also prefer offline verifiable though

Reviewer 3: Resolving as I've merged in the _offline verifiable_ change. Please re-open if folks don't agree.
